src/python/pants/engine/internals/synthetic_targets.py

```diff
@@ -0,0 +1,132 @@
+
+    from pants.engine.internals.synthetic_targets import (
+        SyntheticTargetAdaptor,
+        SyntheticTargetAdaptors,
+        SyntheticTargetAdaptorsRequest,
+    )
+    from pants.engine.unions import UnionRule
+    from pants.engine.rules import collect_rules, rule
+
+
+    class SyntheticExampleTargetAdaptorsRequest(SyntheticTargetAdaptorsRequest):
+        pass
+
+
+    @rule
+    async def example_synthetic_targets(_: SyntheticExampleTargetAdaptorsRequest) -> SyntheticTargetAdaptors:
+        return SyntheticTargetAdaptors(
+            [
+                SyntheticTargetAdaptor.create(
+                    "<spec-path>",
+                    "<target-alias>",
+                    "<target-name>",
+                    **target_kwargs,
+                ),
+                ...
+            ]
+        )
+
+
+    def rules():
+        return (
+            *collect_rules(),
+            UnionRule(SyntheticTargetAdaptorsRequest, SyntheticExampleTargetAdaptorsRequest),
+            ...
+        )
+"""
+from __future__ import annotations
+
+import itertools
+import os.path
+from dataclasses import dataclass
+from typing import cast
+
+from pants.engine.collection import Collection
+from pants.engine.internals.mapper import AddressMap
+from pants.engine.internals.target_adaptor import TargetAdaptor
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.unions import UnionMembership, union
+
+SYNTHETIC_BUILD_FILE_NAME = "-BUILD-synthetic-"
+
+
+@dataclass
+class SyntheticTargetAdaptor:
+    """Synthesizes a target, as if declared in a BUILD file path."""
+
+    path: str
+    target_adaptor: TargetAdaptor
+
+    @classmethod
+    def create(cls, path: str, type_alias: str, name: str, **kwargs) -> SyntheticTargetAdaptor:
+        """The `name` value is required for synthetic targets."""
+        return cls(
+            path=path, target_adaptor=TargetAdaptor(type_alias=type_alias, name=name, **kwargs)
+        )
+
+
+class SyntheticTargetAdaptors(Collection[SyntheticTargetAdaptor]):
+    """A collection of SyntheticTargetAdaptor."""
+
+
+class AllSyntheticTargetAdaptors(SyntheticTargetAdaptors):
+    """All SyntheticTargetAdaptors."""
+
+
+@union
+class SyntheticTargetAdaptorsRequest:
+    pass
+
+
+class SyntheticTargetPaths(Collection[str]):
+    pass
+
+
+@dataclass(frozen=True)
+class SyntheticTargetsAddressMapRequest:
+    path: str
+
+
+@rule
+async def get_all_synthetic_target_paths(
+    all_synthetic_targets: AllSyntheticTargetAdaptors,
+) -> SyntheticTargetPaths:
+    return SyntheticTargetPaths(synthetic_target.path for synthetic_target in all_synthetic_targets)
+
+
+@rule
+async def get_synthetic_target_address_map(
+    request: SyntheticTargetsAddressMapRequest, all_synthetic_targets: AllSyntheticTargetAdaptors
+) -> AddressMap:
+    name_to_target_adaptor: dict[str, TargetAdaptor] = {}
+    for synthetic in all_synthetic_targets:
+        if synthetic.path == request.path:
```

Comment:
This would be O(N*M) for N generated targets and M directories. You could possibly organize `AllSyntheticTargetAdaptors` to be a `dict[path, dict[str, TargetAdaptor]]`? But if you adjust `SyntheticTargetAdaptorsRequest` to be directory-at-a-time, this won't matter.

Reply:
I'm not sure I've not just pushed this `O(N*M)` complexity further out now as each union member now will be called for each directory and looping over all it's possibilities for a match maybe it actually is slightly worse (see #17097 for an example of what I mean), but maybe that's the part that now "doesn't matter"?

Reply:
Thanks for making the change!

> I'm not sure I've not just pushed this `O(N*M)` complexity further out now as each union member now will be called for each directory and looping over all it's possibilities for a match maybe it actually is slightly worse (see #17097 for an example of what I mean), but maybe that's the part that now "doesn't matter"?

Shouldn't this now actually be just a dict-lookup for the name?

Reply:
https://github.com/pantsbuild/pants/pull/17097/files#diff-e15283e30aacb25ae1fa37f0b33fab110d281d2322c064fd2f61c6ba9f40984dR281-R299

Same lines, but outside of a diff window:

https://github.com/kaos/pants/blob/d794dc65634db2c4c8ba48d644f9161076709a29/src/python/pants/backend/python/goals/lockfile.py#L281-L299

bah, was trying to get GH to give me an inline code box..

> Shouldn't this now actually be just a dict-lookup for the name?

You mean `path` ?
As I got rid of the "fetch all synthetic target adaptors" upfront loading rule to instead fetch them per directory I've pushed the overhead of potentially caching the computation onto the union members.

To me it feels more efficient to ask each union member for all its synthetic target adaptors in one go (like `AllSyntheticTargetAdaptors` did) but then do what you suggested to key that off into a per path dictionary that we can use to lookup into later.

Reply:
I think it comes down to whether the files are static and known ahead of time (as in your lockfile case), or whether you actually need to locate them on disk. See my other comment.